bundle/edu.gemini.programexport.servlet/src/main/scala/edu/gemini/programexport/servlet/ProgramExportServlet.scala

```diff
@@ -0,0 +1,209 @@
+  def build(id: SPProgramID, response: HttpServletResponse): Unit =
+    Option(odb.lookupProgramByID(id)) match {
+      case Some(ispProgram) =>
+        val out: BufferedOutputStream = new BufferedOutputStream(response.getOutputStream)
+        response.setContentType("text/plain")
+
+        // TODO: Remove
+        out.write(s"Program name received: $id\n".getBytes)
+
+        val json = process(ispProgram)
+        out.write(json.getBytes)
+        out.close()
+      case None =>
+        val error = s"Program $id is not in the database!"
+        Log.severe(error)
+        throw new ServletException(error)
+    }
+
+  // Recursive method to build up the JSON representation of the program.
+  def process(node: ISPNode): String = {
+    node match {
+      case p: ISPProgram => processProgram(p)
+      case g: ISPGroup => processGroup(g)
+
+      // Other obs component.
+      case n: ISPObsComponent => processObsComponent(n)
+      case _ => ""
+    }
+  }
+
+  // Combine components at the same level into a JSON list.
+  def combine(lst: List[String]): String =
+    ""
+
+  // Top-level program information.
+  def processProgram(ispProgram: ISPProgram): String = {
+    val result = combine(ispProgram.getObsComponents.asScala.map(process).toList)
+
+    // NOTE: Some of these can be null. Indicated for each field.
+    val spProg = ispProgram.getDataObject.asInstanceOf[SPProgram]
+    val gsa = spProg.getGsaAspect
+
+    // Is this all needed? Any other PI info? Not specified in doc.
+    //        val piFirstName = spProg.getPIFirstName
+    //        val piLastName  = spProg.getPILastName
+    //        val piEmail     = spProg.getPIInfo.getEmail // Needed?
+
+    // Investigator data
+    val gsaPhase1Data = spProg.getGsaPhase1Data // not null
+    gsaPhase1Data.getCois.asScala.foreach { inv =>
+      val first = inv.getFirst
+      val last = inv.getLast
+      val email = inv.getEmail
+    }
+
+    val affiliate = spProg.getPIAffiliate.displayValue // could be null
+    val queueBand = spProg.getQueueBand // String, empty if classical
+    val rolloverFlag = spProg.getRolloverStatus // boolean
+    val isThesis = spProg.isThesis // boolean
+    val programMode = spProg.getProgramMode // isQueue / isClassical // not null
+    val tooType = spProg.getTooType.getDisplayValue // not null
+    // TODO: In what format?
+    val awardedTime = spProg.getAwardedProgramTime // TimeValue, not null
+
+    // All non-null
+    val timeAcctAllocation = spProg.getTimeAcctAllocation
+    val timeAcctCategories = timeAcctAllocation.getCategories
+    timeAcctCategories.asScala.foreach { category =>
+      // Times are Durations. Bryan wants int seconds or nights if nights are specified.
+      // Not sure how possible this is?
+      val timeAcctAward = timeAcctAllocation.getAward(category)
+      val programTime = timeAcctAward.getProgramAward
+      val partnerTime = timeAcctAward.getPartnerAward
+    }
+
+    ""
```

Comment:
To be done?

Reply:
Yup... that is where the JSON (however it will be built) was to be put together as I had the data I needed.
Shane is going to help me work on a way to traverse the data model better so that this task is easier, since the nesting will happen naturally. I just can't figure out the matching on my own.